streams/src/test/java/org/apache/kafka/streams/kstream/internals/KTableSourceTest.java

```diff
@@ -85,6 +85,36 @@ public void testKTable() {
             supplier.theCapturedProcessor().processed);
     }
 
+    @Test
+    public void testKTableEmitOnChange() {
+        final StreamsBuilder builder = new StreamsBuilder();
+        final String topic1 = "topic1";
+
+        final KTable<String, Integer> table1 =
+            builder.table(topic1, Consumed.with(Serdes.String(), Serdes.Integer()), Materialized.as("store"));
+
+        final MockProcessorSupplier<String, Integer> supplier = new MockProcessorSupplier<>();
+        table1.toStream().process(supplier);
+
+        try (final TopologyTestDriver driver = new TopologyTestDriver(builder.build(), props)) {
+            final TestInputTopic<String, Integer> inputTopic =
+                    driver.createInputTopic(topic1, new StringSerializer(), new IntegerSerializer());
+            inputTopic.pipeInput("A", 1, 10L);
+            inputTopic.pipeInput("B", 2, 11L);
+            inputTopic.pipeInput("A", 1, 10L);
+            inputTopic.pipeInput("B", 3, 13L);
+
+            assertEquals(1.0,
+                getMetricByName(driver.metrics(), "idempotent-update-skip-total", "stream-processor-node-metrics").metricValue());
+        }
+
+        assertEquals(
+            asList(new KeyValueTimestamp<>("A", 1, 10L),
+                new KeyValueTimestamp<>("B", 2, 11L),
+                new KeyValueTimestamp<>("B", 3, 13L)),
+            supplier.theCapturedProcessor().processed);
+    }
```

Comment:
```suggestion
    @Test
    public void testKTableEmitOnChange() {
        final StreamsBuilder builder = new StreamsBuilder();
        final String topic1 = "topic1";

        builder.table(topic1, Consumed.with(Serdes.String(), Serdes.Integer()), Materialized.as("store"))
               .toStream()
               .to("output");

        try (final TopologyTestDriver driver = new TopologyTestDriver(builder.build(), props)) {
            final TestInputTopic<String, Integer> inputTopic =
                driver.createInputTopic(topic1, new StringSerializer(), new IntegerSerializer());
            final TestOutputTopic<String, Integer> outputTopic =
                driver.createOutputTopic("output", new StringDeserializer(), new IntegerDeserializer());

            inputTopic.pipeInput("A", 1, 10L);
            inputTopic.pipeInput("B", 2, 11L);
            inputTopic.pipeInput("A", 1, 12L);
            inputTopic.pipeInput("B", 3, 13L);

            assertThat(
                getMetricByName(driver.metrics(), "idempotent-update-skip-total", "stream-processor-node-metrics").metricValue(),
                is(1.0)
            );

            assertThat(
                outputTopic.readRecordsToList(),
                is(
                    asList(new TestRecord<>("A", 1, Instant.ofEpochMilli(10L)),
                           new TestRecord<>("B", 2, Instant.ofEpochMilli(11L)),
                           new TestRecord<>("B", 3, Instant.ofEpochMilli(13L)))
                )
            );
        }
    }
```

Very nice test!

Two suggestions for the price of one:
1. It's basically the same either way, but since we're using TopologyTestDriver and TestInputTopic, let's use TestOutputTopic as well instead of MockProcessorSupplier for the verification.
2. In the KTable context, we agreed to consider the update idempotent even if the timestamp is different, as long as the value is the same. Therefore, I've changed the update to `A` to have timestamp 12 instead of 10.

Of course, we'd need several more tests to really verify all the angles of this feature.